### PR TITLE
Make value and defaultValue valid in number and slider fields

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/4978
Closes https://github.com/adobe/react-spectrum/issues/5556

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-spectrum/numberfield/test/NumberField.test.js [hunk 1]
```diff
@@ -295,6 +295,21 @@ describe('NumberField', function () {
...
 
+  it.each`
+  Name                            | props                                        | expected
+  ${'defaultValue minValue'}      | ${{defaultValue: 20, minValue: 50}}          | ${'50'}
+  ${'defaultValue maxValue'}      | ${{defaultValue: 20, maxValue: 10}}          | ${'10'}
+  ${'defaultValue minValue step'} | ${{defaultValue: 20, minValue: 50, step: 3}} | ${'50'}
+  ${'defaultValue maxValue step'} | ${{defaultValue: 20, maxValue: 10, step: 3}} | ${'9'}
+  ${'value minValue'}             | ${{value: 20, minValue: 50}}                 | ${'50'}
+  ${'value maxValue'}             | ${{value: 20, maxValue: 10}}                 | ${'10'}
+  ${'value minValue step'}        | ${{value: 20, minValue: 50, step: 3}}        | ${'50'}
```
Comment (ktabors) [inline packages/@react-spectrum/numberfield/test/NumberField.test.js:306]: I've forgotten why this doesn't clamp to 51. Is it that the step interval start at zero or min?

**Review by ktabors — APPROVED** [review]
Should we add a story that connects a NumberField and Slider with the Slider bound to a min/max range and the NumberField allows any number so we can give it values outside the range? RAC Slider docs have an example that does this so we are probably fine.

File: packages/@react-spectrum/numberfield/test/NumberField.test.js (hunk 1 above)
Comment (snowystinger) [reply packages/@react-spectrum/numberfield/test/NumberField.test.js:306]: steps start at the minValue

**Comment by snowystinger** [comment]
> Should we add a story that connects a NumberField and Slider with the Slider bound to a min/max range and the NumberField allows any number so we can give it values outside the range? RAC Slider docs have an example that does this so we are probably fine.

If RAC has it, we're fine, also we have tests for this now

File: packages/@react-stately/numberfield/src/useNumberFieldState.ts [hunk 2]
```diff
@@ -90,6 +90,26 @@ export function useNumberFieldState(
...
+    value = NaN;
+  }
+
+  if (!isNaN(value)) {
+    if (!isNaN(step)) {
+      value = snapValueToStep(value, minValue, maxValue, step);
+    } else {
+      value = clamp(value, minValue, maxValue);
+    }
+  }
```
Comment (LFDanLu) [inline packages/@react-stately/numberfield/src/useNumberFieldState.ts:103]: so this is a departure from native behavior where the initial value provided to the input is respected even if it doesn't match the step but I saw in the spec where it recommends this new behavior so I'm fine with this, feels better anyways

**Review by LFDanLu — APPROVED** [review]

**Pull request merged**